### PR TITLE
Add RectangularPrism and HexagonalPrism composite surfaces

### PR DESCRIPTION
**Pull request by paulromano**
# Description

One thing in our Python API that has always been confusing to users is the `rectangular_prism` and `hexagonal_prism` functions. As opposed to surfaces (where you need to take `-` or `+` to get a proper region that can be assigned to a cell), these functions return `Region` objects directly, so you can't do:
```Python
prism = openmc.model.rectangular_prism(...)
cell = openmc.Cell(region=-prism)  # The '-' operator doesn't work!
```

This PR proposes two new classes, `RectangularPrism` and `HexagonalPrism`, that are meant to replace those two functions are give us composite surfaces that work with the unary `-` and `+` operators.

# Checklist

- [x] I have performed a self-review of my own code
- [x] <s>I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)</s>
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by paulromano** [comment]
Making a note to myself that this will need to be updated if #2724 is merged first.

File: tests/regression_tests/periodic_hex/test.py [hunk 1]
```diff
@@ -12,8 +12,8 @@ def hex_model():
     fuel.add_nuclide('U235', 1.0)
     fuel.set_density('g/cc', 4.5)
 
-    hex_region = openmc.model.hexagonal_prism(10.0, boundary_type='periodic')
-    cell = openmc.Cell(fill=fuel, region=hex_region)
+    hex_region = openmc.model.HexagonalPrism(10.0, boundary_type='periodic')
+    cell = openmc.Cell(fill=fuel, region=-hex_region)
```
Comment (pshriwise) [inline tests/regression_tests/periodic_hex/test.py:16]:
```suggestion
    hex_prism = openmc.model.HexagonalPrism(10.0, boundary_type='periodic')
    cell = openmc.Cell(fill=fuel, region=-hex_prism)
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
One tiny suggestion but I think this looks good otherwise @paulromano. Thanks!!

**Comment by paulromano** [comment]
Thanks @pshriwise. Just incorporated your suggestion and also resolved a merge conflict.

**Review by pshriwise — APPROVED** [review]

**Pull request merged**